### PR TITLE
BUG: need better inference for path in Series construction (GH9456)

### LINKED ISSUE #9456: BUG: need better inference for path in Series construction
This hits a path in ``Series.__init__`` which I think needs some better inference

https://github.com/pydata/pandas/blob/master/pandas/core/series.py#L178

```
In [1]: d = {numpy.datetime64('2015-01-07T02:00:00.000000000+0200'): 42544017.198965244,
   ...:      numpy.datetime64('2015-01-08T02:00:00.000000000+0200'): 40512335.181958228,
   ...:      numpy.datetime64('2015-01-09T02:00:00.000000000+0200'): 39712952.781494237,
   ...:      numpy.datetime64('2015-01-12T02:00:00.000000000+0200'): 39002721.453793451}

In [2]: Series(d)
Out[2]: 
2015-01-07   NaN
2015-01-08   NaN
2015-01-09   NaN
2015-01-12   NaN
dtype: float64

In [3]: Series(d.values(),d.keys())
Out[3]: 
2015-01-07    42544017.198965
2015-01-08    40512335.181958
2015-01-09    39712952.781494
2015-01-12    39002721.453793
dtype: float64
```

The problem is the index is already converted at this point and its not easy to get the keys/values out (except to do so explicity which is better IMHO).

Need a review of what currently hits this path (can simply put a halt in here and see what tests hit this). Then figure out a better method.

### PR DESCRIPTION
**Pull request by patrickfournier**
closes  #9456.

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -166,9 +166,9 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
                     else:
                         index = Index(_try_sort(data))
                 try:
-                    if isinstance(index, DatetimeIndex):
+                    if isinstance(index, DatetimeIndex) and lib.infer_dtype(data) != 'datetime64':
```
Comment (jreback) [inline pandas/core/series.py]: see why this path is hit in the first place and can go from there.

Comment (patrickfournier) [inline pandas/core/series.py]: Before I added `.value` at line 171, `lib.fast_multiget` was throwing a `TypeError`. Now, the `else` clause must handles the conversion of datetime64 dicts because we cannot compare a `Timestamp` with a `datetime64`:

```python
>>> d = {datetime.datetime(2015, 1, 7, 2, 0): 42544017.198965244}
>>> i = pandas.core.index.Index(d)
>>> i.astype('O').values
array([Timestamp('2015-01-07 02:00:00')], dtype=object)
>>> datetime.datetime(2015, 1, 7, 2, 0) == pandas.lib.Timestamp('2015-01-07 02:00:00')
True

>>> d = {numpy.datetime64('2015-01-06T19:00:00.000000000-0500'): 42544017.198965244}
>>> i = pandas.core.index.Index(d)
>>> i.astype('O').values
array([Timestamp('2015-01-07 00:00:00')], dtype=object)
>>> numpy.datetime64('2015-01-06T19:00:00.000000000-0500') == pandas.lib.Timestamp('2015-01-07 00:00:00')
False
```

Comment (jreback) [inline pandas/core/series.py]: that's not what I mean. put halt in there and see what test actually hits this. This is handling an edge case. Your fix, just confuses things. I think there is a more general soln.

Comment (patrickfournier) [inline pandas/core/series.py]: I ran the tests and reported the results in the issue: https://github.com/pydata/pandas/issues/9456.

Maybe a better fix would be to correctly handle the comparison between Timestamp and datetime64?

Comment (jreback) [inline pandas/core/series.py]: The idea is to have the least special cases as possible. You rarely actually want to convert a ``DatetimeIndex`` using ``.astype('O')`` this is the least desirable result (though it may have been done for some reason). But it is tricky because a user can pass ``Timestamp/datetime.datetime/datetime.date/np.datetime64`` as elements and a non-index like (but that is index-like). Best best is to do ``_ensure_index`` which will convert list-like things to an ``Index`` (and possibly to a ``DatetimeIndex`` if its compat; if its not then all bets are off).

**Comment by patrickfournier** [comment]
I rewrote the fix to be more generic. If the type of Index elements is the same as the type of data elements, we use the Index to access and copy the dict value. If not, we use the original index. See comments in diff.

File: pandas/core/series.py [hunk 2]
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
...
                 try:
-                    if isinstance(index, DatetimeIndex):
-                        # coerce back to datetime objects for lookup
-                        data = lib.fast_multiget(data, index.astype('O'),
-                                                 default=np.nan)
-                    elif isinstance(index, PeriodIndex):
-                        data = [data.get(i, nan) for i in index]
+                    # lib.fast_multiget raises TypeError if type(data) != dict
+
+                    if lib.infer_dtype(data) == lib.infer_dtype(index.values):
```
Comment (patrickfournier) [inline pandas/core/series.py]: Here we try to avoid a call to `np.array()` by checking if we can access `data` values using `index.values`. However I checked the code for `infer_dtype` and it may be quite complex. Maybe we should just go straight to the `else`?

File: pandas/core/series.py [hunk 3]
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
             elif isinstance(data, dict):
                 if index is None:
                     if isinstance(data, OrderedDict):
-                        index = Index(data)
+                        original_index = data.keys()
                     else:
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
+                    index = Index(original_index)
```
Comment (patrickfournier) [inline pandas/core/series.py]: I wanted to use `_ensure_index`, but it does not create a `MultiIndex` when the dict keys are tuples.

File: pandas/core/series.py [hunk 4]
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
...
-                        data = lib.fast_multiget(data, index.values,
-                                                 default=np.nan)
-                except TypeError:
-                    data = [data.get(i, nan) for i in index]
+                        if isinstance(original_index, PeriodIndex):
+                            data = [data.get(i, np.nan) for i in original_index]
+                        else:
+                            # np.array(['z', ('a', 'b')]) raises ValueError;
+                            # this may happens with MultiIndex.
+                            data = lib.fast_multiget(data, np.array(original_index),
```
Comment (patrickfournier) [inline pandas/core/series.py]: If we `lib.fast_multiget` supported `list` as its second parameter, we could save the call to `np.array`.

File: pandas/core/series.py [hunk 5]
```diff
@@ -162,21 +164,27 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
             elif isinstance(data, dict):
                 if index is None:
                     if isinstance(data, OrderedDict):
-                        index = Index(data)
+                        original_index = data.keys()
                     else:
-                        index = Index(_try_sort(data))
+                        original_index = _try_sort(data)
+
```
Comment (jreback) [inline pandas/core/series.py]: why is it necessary to use a new variable here. This original code looks fine.

Comment (patrickfournier) [inline pandas/core/series.py]: Since we cannot always reorder the `dict` using the `Index` (because of incompatible data types), I keep the original index to use it if I cannot use the `Index`.

**Comment by jreback** [comment]
this overlaps with #10269

**Comment by jreback** [comment]
superseded by #10269 

thanks!
